File: app/src/main/java/net/mmho/photomap2/ThumbnailFragment.java
package net.mmho.photomap2;

import android.content.Intent;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.os.Parcelable;
import android.support.v4.app.Fragment;
import android.support.v4.util.LruCache;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.GridView;

public class ThumbnailFragment extends Fragment {

    private PhotoGroup group;
    private GridView list;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);
        setHasOptionsMenu(true);
    }

    public void setList(PhotoGroup g){
        group = g;
        ThumbnailAdapter adapter =
            new ThumbnailAdapter(getActivity(), R.layout.adapter_thumbnail, group);
        list.setAdapter(adapter);
    }


    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        String title = group.getTitle();
        if(title!=null){
            getActivity().setTitle(title);
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View parent = inflater.inflate(R.layout.fragment_thumbnail,container,false);
        list = (GridView)parent.findViewById(R.id.thumbnail_grid);
        list.setOnItemClickListener(clickListener);
        return parent;
    }

    public void setPosition(int position){
        list.setSelection(position);

    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.thumbnail_manu,menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch(item.getItemId()){
            case R.id.map:
                Intent i = new Intent(getActivity(),PhotoMapActivity.class);
                i.putExtra(PhotoMapFragment.EXTRA_GROUP, (Parcelable) group);
                startActivity(i);
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    AdapterView.OnItemClickListener clickListener =
        new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                Intent i = new Intent(getActivity(),PhotoViewActivity.class);
                i.putExtra(PhotoViewActivity.EXTRA_GROUP, (Parcelable) group);
                i.putExtra(PhotoViewActivity.EXTRA_POSITION,position);
                startActivityForResult(i,0);
            }
        };
}

File: app/src/main/kotlin/net/mmho/photomap2/ThumbnailFragment.kt
package net.mmho.photomap2

import android.content.Intent
import android.os.Bundle
import android.os.Parcelable
import android.support.v4.app.Fragment
import android.view.*
import android.widget.AdapterView
import android.widget.GridView
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.layout_thumbnail.view.*

class ThumbnailFragment : Fragment() {

    private var group: PhotoGroup? = null
    private var list: GridView? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        retainInstance = true
        setHasOptionsMenu(true)
    }

    fun setList(g: PhotoGroup) {
        group = g
        val adapter = ThumbnailAdapter(requireActivity(), R.layout.adapter_thumbnail, group as PhotoGroup)
        list?.adapter = adapter
    }


    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        Observable
            .just(group)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .doOnNext { g -> g?.resolveAddress(requireContext())}
            .doOnComplete { requireActivity().title = group?.title }
            .subscribe()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val parent = inflater.inflate(R.layout.fragment_thumbnail, container, false)
        list = parent.thumbnail_grid
        list?.onItemClickListener = clickListener
        return parent
    }

    fun setPosition(position: Int) {
        list?.setSelection(position)
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater) {
        inflater.inflate(R.menu.thumbnail_manu, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item?.itemId) {
            R.id.map -> {
                val i = Intent(activity, PhotoMapActivity::class.java)
                i.putExtra(PhotoMapFragment.EXTRA_GROUP, group as Parcelable?)
                startActivity(i)
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        val position = data?.getIntExtra(PhotoViewActivity.EXTRA_POSITION,0) ?: 0
        setPosition(position)
    }

    private val clickListener = AdapterView.OnItemClickListener { _, _, position, _ ->
        val i = Intent(activity, PhotoViewActivity::class.java)
        i.putExtra(PhotoViewActivity.EXTRA_GROUP, group as Parcelable?)
        i.putExtra(PhotoViewActivity.EXTRA_POSITION, position)
        startActivityForResult(i, 0)
    }
}
